File: archive/p/python/factorial.py
import sys


def factorial(n):
    if n <= 0:
        return 1
    return n * factorial(n - 1)


def exit_with_error(msg=None):
    msg = msg or 'Usage: please input a non-negative integer'
    print(msg)
    sys.exit(1)


def main(args):
    try:
        n = int(args[0])
        if n < 0:
            exit_with_error()
        elif n >= 996:
            msg = f'{n}! is out of the reasonable bounds for calculation'
            exit_with_error(msg)
        print(factorial(n))
    except (IndexError, ValueError):
        exit_with_error()


if __name__ == "__main__":
    main(sys.argv[1:])

File: archive/p/pascal/factorial.pas
program factorial_calculate(input, output, stdErr);
(* Read a number from Commandline, Return it's factorial value
Cardinal, int-64 and qWord are the relevant data type to store long int in pascal.
With Cardinal data type, maximum factorial that can be computed is 34, 4926277576697053184
With int-64   data type, maximum factorial that can be computed is 44, 2673996885588443136
With qWord    data type, maximum factorial that can be computed is 65, 9223372036854775808
*)
var
 buf: String;
(* input_value, factorial, check:qWord;*)
i, check: integer;
input_value, factorial_value: qWord;
begin    
  buf:= paramStr(1);
  Val(buf, input_value, check);  
  if check <> 0
  then
    writeln('Usage: please input a non-negative integer')
  else
    if input_value < 0
      then
        writeln('Usage: please input a non-negative integer')
      else
      begin
        if (input_value = 0) then 
          writeln('1')
        else
          begin
            factorial_value := 1;
            for i := 1 to input_value do 
              begin
              factorial_value := factorial_value * i;
              end;
            writeln(factorial_value);
          end;

      end;
end.
